Linux app

benkolera opened the pull request:
This both builds a nixos app and a debian. 

To run nixos:
`WEBKIT_DISABLE_COMPOSITING_MODE=1 $(nix-build -A linux)/bin/kadena-chainweaver` from the source

To run debian:
- In this source directory `nix-build -A deb`
- copy result/chainweaver_0.1.0_amd64.deb to an ubuntu machine. 
- On ubuntu machine terminal: 
  - `sudo dpkg -i kadena-chainweaver_0.1.0_amd64.deb`
  - Either run it from the application menu or run `kadena-chainweb-rc1` in a terminal if something is bad and you need to see some logs.

Current known issues across both nixos and debian:
- [x] Pact file not saved across executions of chainweaver
- [ ] Warnings about ArrayBuffer vs ArrayBufferView in the console
- [x] Doesn't output to syslog and just the console atm
- [x] Problems with compositing with nvidia and in virtual box. Best to disable
- [x] Writes to ~/.local/shared/chainweaver instead of .local/share. Typo

Current known issues for nixos:
- Doesn't disable compositing automatically in the wrapper script

Current known issues for debian:
- [x] First run is painfully slow (come up as not responding; may be my vm only has 4G mem)
- [x] External links don't open
- [x] File dialog crashes app
- [x] .deb could be smaller
- [x] Put static files in a better place than libexec
- [x] Signing server not running / responding
- [x] Needs to package up a pixbuf config and maybe fix gsettings provider too
- [x] Most likely want a desktop entry rather than running from a terminal :) 
- [x] With no CI / binary caching atm, the custom rebuild of webkitgtk is painful because it takes about 40 mins to build. 
- [ ] No testing on Mate (GTK 2), Cinnamon, XCFE distros. All testing has been done on Ubuntu 18 (GTK 3 and XOrg). Given that the deb is so self contained, it's probably more wayland that will mess us up rather than non-gtk3 distros. 

Alternative ways of going about this:
- Instead of copying the libraries across, statically link. Involves compiling things with musl instead which is probably more trouble than it's worth. Given that GTK is lgpl, its in our best interests not to statically link to it anyway. See: https://github.com/nh2/static-haskell-nix/issues/50
- Something like https://github.com/matthewbauer/nix-bundle where we package up a whole nix store . This produced a thing that was quite large and out of the box had a pretty slow startup time.

benkolera committed:
first go

benkolera committed:
Merge remote-tracking branch 'origin/redesign' into blk@redesign/linux-app

benkolera committed:
feat: First go at a linux app

benkolera committed:
feat: comment about running it

benkolera committed:
feat: Add comment about upgrading gtk

benkolera committed:
Merge remote-tracking branch 'origin/redesign' into blk@redesign/linux-app

benkolera committed:
fix: more notes

benkolera committed:
Merge remote-tracking branch 'origin/redesign' into blk@redesign/linux-app

benkolera committed:
feat: Working file dialog, focus, iconify

benkolera committed:
feat: Some extra notes

benkolera committed:
Moving towards a .deb

alexfmpe commented on default.nix:
```diff
@@ -103,6 +104,91 @@ in obApp // rec {
+      cp -rL "${obApp.passthru.staticFiles}" "$BINDIR/static"
+      cp -rL "${sass}/sass.css" "$BINDIR/sass.css"
+      ${pkgs.libicns}/bin/png2icns "$BINDIR/pact.icns" "${macAppIcon}"
+      ${pkgs.libicns}/bin/png2icns "$BINDIR/pact-document.icns" "${macPactDocumentIcon}"
```
Don't know how linux app icons work, but I doubt they use Apple Icon Image format

benkolera committed:
OK, this boots in ubuntu. File dialog segfaults ubuntu and webkit inspector doesn't open (these appear related). Network calls broken in nixos and ubuntu.

benkolera committed:
Making a nice looking deb now

Problems still remaining:
- Should package up a pixbuf cache
- Trim down unused shared objects
- Can't open external links
- File dialog still segfaults the app
- ArrayBuffer vs ArrayBufferView warnings

Problems fixed:
- Can open up inspector now
- Disables compositing, includes glib-networking, wrapper script

benkolera committed:
Fixed some notes

benkolera committed:
Merge remote-tracking branch 'origin/redesign' into blk@redesign/linux-app

benkolera committed:
Removed the manual wrapper script

matthewbauer commented on default.nix:
```diff
@@ -103,6 +104,203 @@ in obApp // rec {
+      cp $TMPEXE $CHAINWEAVER_LIBEXEC_EXE
+      cp "${pkgs.z3}"/bin/z3 "$LIBEXECDIR/z3"
+
+      #TODO : This shouldn't be in libexec
```
Yeah, obelisk by default looks in executable directory for resources. You can use something like this to make it configurable via a cli flag:

- https://github.com/obsidiansystems/obelisk/pull/558
- https://gitlab.com/obsidian.systems/kiln/merge_requests/733/diffs#diff-content-5b1d1ec64e71b88e93f64a6e7dbb173c49239c6e

matthewbauer commented on default.nix:
```diff
@@ -103,6 +104,203 @@ in obApp // rec {
+      # Figure out how to make the array work with nix escaping
+      wrapperArgs=""
+      wrapperArgs+="--set WEBKIT_DISABLE_COMPOSITING_MODE 1 "
+      wrapperArgs+="--set LD_LIBRARY_PATH $LIBPATH "
```
probably should be prefix to let others also override other things.

benkolera committed:
Copy gsetting across

benkolera committed:
Merge remote-tracking branch 'origin/develop' into blk@redesign/linux-app

benkolera committed:
File dialogs work and external links work

benkolera committed:
Merge remote-tracking branch 'origin/develop' into blk@redesign/linux-app

benkolera committed:
Desktop item

benkolera committed:
Tidy some cruft up

benkolera committed:
Merge branch 'blk@ts-split-accounts/versioning' into blk@redesign/linux-app

benkolera committed:
Merge remote-tracking branch 'origin/ts-split-accounts' into blk@redesign/linux-app

Commit:
WIP created logging functions, replaced some to get a feel for the utility of it

benkolera committed:
Loads initial editor file from storage

benkolera committed:
feat: Kadena blue scrollbars with a less paperthin width in webkit

Commit:
Replace putStrLn with logging calls

Commit:
Pass whole logger around instead of individual functions, this is more consistent with passing around the 'Network' info

Commit:
Rename logPromptly to putLog

benkolera committed:
refactor: Split default into linux and mac files

Commit:
Add hsyslog package

benkolera committed:
Add missing Mac.hs imports

benkolera committed:
s/MacFFI/AppFFI/

benkolera committed:
Merge branch 'blk@redesign/linux-app' of github.com:kadena-io/chainweaver into blk@redesign/linux-app

benkolera committed:
Merge remote-tracking branch 'origin/develop' into blk@redesign/linux-app

benkolera committed:
Merge branch 'blk@redesign/linux-app' of github.com:kadena-io/chainweaver into blk@redesign/linux-app

benkolera committed:
macFFI -> appFFI

benkolera committed:
Inherit deb, not exe

Commit:
Add syslog to linux & mac builds

benkolera committed:
Final linux fixes

benkolera committed:
Merge branch 'blk@redesign/linux-app' into sc-add-logging-functions

benkolera committed:
Merge pull request #376 from kadena-io/sc-add-logging-functions

Centralise the logging & remove all `putStrLn` calls

benkolera committed:
Missed some print lines

benkolera committed:
Changelog

tomsmalley commented on linux/src/Linux.hs:
```diff
@@ -0,0 +1,183 @@
+{-# LANGUAGE OverloadedStrings #-}
+module Linux where
+
+-- import Control.Lens (_head, preview)
```
leftover

tomsmalley commented on linux/src/Linux.hs:
```diff
@@ -0,0 +1,183 @@
+To run: `WEBKIT_DISABLE_COMPOSITING_MODE=1 $(nix-build -A linux)/bin/kadena-chainweaver-rc1` from the project root
+
+TODO
+- Setting up app menu? What does the mac app do and what do we want in there?
```
It just allows opening a file, quitting, and it also enables the copy/paste shortcuts. Probably not applicable here

benkolera committed:
Merge branch 'develop' into blk@redesign/linux-app

tomsmalley commented on linux/src/Linux.hs:
```diff
@@ -0,0 +1,183 @@
+
+moveToForeground :: Gtk.Window -> IO ()
+moveToForeground w = do
+  putStrLn "MoveToForeground"
```
can these use logToSyslog?

benkolera replied:
I don't think they are even useful anymore. I'll just nuke them. :)

tomsmalley reviewed (approved):
nitpicks only

benkolera committed:
Syslog commonisation. Mac fixes

benkolera committed:
Syslog commonisation. Mac fixes

benkolera committed:
Merge branch 'blk@redesign/linux-app' of github.com:kadena-io/chainweaver into blk@redesign/linux-app

benkolera committed:
Mac fixes

benkolera committed:
Merge remote-tracking branch 'origin/develop' into blk@redesign/linux-app
